File: track.py
import os
from random import random
import urllib.request

from mutagen import id3

cache = os.getenv('YANDEX_MUSIC_CACHE', os.getenv('TMPDIR'))
assert cache, 'CACHE DIR NOT DEFINED'

class Track:
    def __init__(self, track, station):
        self.station = station
        self.id = track['id']
        self.album_id = track['albums'][0]['id']
        self.play_id = "%s-%s-%s" % (
            int(random() * 1000),
            int(random() * 1000),
            int(random() * 1000)
        )
        self.duration = int(track['durationMs'] / 1000)
        self.position = 0

        self.artist = track['artists'][0]['name']
        self.album = track['albums'][0]['title']
        self.title = track['title']

        self.name = f"{self.artist} - {self.title}".replace("/", "")
        self.path = f'{cache}/{self.name}.mp3'

        self._lyrics = None
        self._lyrics_available = track['lyricsAvailable']

        self.cached = self.exists
        self.artwork = f'https://{track["coverUri"].replace("%%", "200x200")}'

        self.artists = " feat. ".join([a['name'] for a in track['artists']])

    @property
    def lyrics(self):
        if not self._lyrics:
            if self._lyrics_available:
                self._lyrics = self.station.client.get_lyrics(self.id)

        return self._lyrics

    @property
    def exists(self):
        return os.path.exists(self.path)

    def download(self):
        if not self.exists:
            print(f'Downloading {self.name}...')
            os.makedirs("/".join(self.path.split("/")[:-1]), exist_ok=True)
            self.station.client.download(self.id, self.path)

        self.__reload_tags()

    def trace(self):
        self.station.event_track_trace(self)

    def skip(self):
        self.station.event_track_skip(self)

    def like(self):
        self.station.event_track_like(self)

    def dislike(self):
        self.station.event_track_dislike(self)
        if self.exists:
            os.remove(self.path)

    def __reload_tags(self):
        if not self.exists:
            return
        try:
            tags = id3.ID3(self.path)
        except id3.ID3NoHeaderError:
            tags = id3.ID3()

        changed = False

        if not tags.get('TIT2'):
            tags.add(id3.TIT2(encoding=3, text=self.title))
            tags.add(id3.TALB(encoding=3, text=self.album))
            tags.add(id3.TPE1(encoding=3, text=self.artist))
            print("Title changed")
            changed = True

        if tags.get('USLT::eng'):
            self._lyrics = tags['USLT::eng'].text
        elif self.lyrics is not None:
            tags.add(id3.USLT(encoding=3, lang='eng', text=self.lyrics))
            print("Lyrics changed")
            changed = True

        if not tags.get('APIC'):
            art = urllib.request.urlopen(self.artwork)
            tags.add(id3.APIC(
                encoding=0,
                mime=art.headers['Content-Type'],
                type=3,  # cover front
                data=art.read(),
            ))
            print("Artwork changed")
            changed = True

        if changed:
            tags.save(self.path)

File: OlaRadioUI/Models/Track.swift
//
//  Track.swift
//  OlaRadioUI
//
//  Created by Mike Miklin on 14.11.2022.
//

import Foundation

final public class Track: ObservableObject {
    let packet: TrackPacket
    let station: Station

    let id: String
    let title: String
    var liked = false
    var cached = false
    @Published private var _lyrics: String? = nil
    private var lyricsAvailable: Bool = false
    var task: Task<(), Error>?
    var cache: String {
        UserDefaults.standard.string(forKey: "cache")
            ?? ProcessInfo.processInfo.environment["TMPDIR"] ?? "/tmp"
    }

    let playId = String(
        format: "%d-%d-%d",
        Int.random(in: 1...1000),
        Int.random(in: 1...1000),
        Int.random(in: 1...1000)
    )
    var album: String { packet.albums.first!.title }
    var albumId: Int { packet.albums.first!.id }
    var artist: String { packet.artists.first!.name }
    var name: String { "\(artist) - \(title)".replacingOccurrences(of: "/", with: "") }
    var path: String { "\(cache)/\(name).mp3" }
    var artwork: String {
        "https://\(packet.coverUri.replacingOccurrences(of: "%%", with: "200x200"))"
    }
    private static let sessionProcessingQueue = DispatchQueue(label: "SessionProcessingQueue")

    var duration: Int { packet.durationMs / 1000 }
    var position: Int = 0
    var exists: Bool {
        FileManager.default.fileExists(atPath: path)
    }

    var lyrics: String? {
        get {
            if _lyrics == nil && lyricsAvailable {
                Task { try await getLyrics() }
            }
            return _lyrics
        }
        set {
            _lyrics = newValue
        }
    }
    func getLyrics() async throws {
        let lyrics = try await station.client.getLyrics(trackId: id)
        DispatchQueue.main.sync {
            self.lyrics = lyrics
        }
    }

    init(_ packet: TrackPacket, _ station: Station) {
        self.station = station
        self.packet = packet
        id = packet.id
        title = packet.title
        liked = packet.liked
        lyricsAvailable = packet.lyricsAvailable
        cached = exists
    }

    func download() async throws {
        if !exists {
            print("Downloading \(name)...")
            try FileManager.default.createDirectory(
                atPath: path.split(separator: "/").dropLast().joined(separator: "/"),
                withIntermediateDirectories: true
            )
            try await station.client.download(trackId: id, filename: path)
        }

        // self.__reload_tags()
    }

    func trace() async throws {
        try await station.eventTrackTrace(track: self)
    }

    func skip() async throws {
        try await station.eventTrackSkip(track: self)
    }

    func like() async throws {
        try await station.eventTrackLike(track: self, remove: liked)
        DispatchQueue.main.sync {
            self.liked.toggle()
        }
    }

    func dislike() async throws {
        try await station.eventTrackDislike(track: self)
        try FileManager.default.removeItem(atPath: path)
    }
}
